docs: guide for migrating to publishable and secret API keys (#46600)
supabase/supabase: apps/docs/content/guides/getting-started/migrating-to-new-api-keys.mdx

```ts
import { createClient } from '@supabase/supabase-js'

const supabase = createClient(
  'https://your-project.supabase.co',
  'sb_publishable_...' // was the anon key
)
```

The publishable key carries the same low privileges as the `anon` key, so your [Row Level Security](/docs/guides/database/postgres/row-level-security) policies behave the same. User authentication through Supabase Auth is unchanged: the user still signs in and gets their own JWT.

## Step 3: Swap the secret key in backend code

Anywhere you use the `service_role` key on a server you control, switch to a secret key.

```ts
import { createClient } from '@supabase/supabase-js'

const supabaseAdmin = createClient(
  'https://your-project.supabase.co',
  'sb_secret_...' // was the service_role key
)
```

Secret keys add protections the `service_role` key doesn't have. They return HTTP 401 if used in a browser (matched on the `User-Agent` header), and you can run a separate key per service so a single leak only forces one rotation.

<Admonition type="caution">

> Secret keys bypass Row Level Security and have full access to your data. Keep them on backends you control, out of source control, and out of client code.

Author: Plural seems right to me in this context

</Admonition>

> ### Database Webhooks and `pg_net`

Author: Wondering if this applies when referring to the product `Database Webhooks` (https://supabase.com/dashboard/project/_/integrations/webhooks/overview)

> ### Database Webhooks and `pg_net`

Reviewer: We don't enforce our own style guide everywhere 😆. But I will add this as an exception.

---
id: 'migrating-to-new-api-keys'

> title: 'Migrate to publishable and secret API keys'

Reviewer: ```suggestion
title: 'Migrating to publishable and secret API keys'
```

> description: 'Move from legacy JWT-based anon and service_role keys to publishable and secret keys with zero downtime.'

Reviewer: ```suggestion
description: 'Move from legacy JWT-based anon and service_role keys to publishable and secret keys.'
```

Let's not over promise

---

Supabase has changed the way API keys work. The legacy `anon` and `service_role` keys are based on your project's JWT secret, which makes them hard to rotate without downtime. The new publishable (`sb_publishable_...`) and secret (`sb_secret_...`) keys can be created, named, and revoked independently, so you can rotate a single key without touching the rest of your app.

> This guide walks through migrating an existing project. Both key types work at the same time, so you can swap clients one at a time and deactivate the legacy keys only once nothing depends on them.

Reviewer: ```suggestion
This guide covers migrating an **existing project.** Both key types work simultaneously, so you can swap clients one at a time and deactivate the legacy keys only after nothing depends on them.
```

<Admonition type="note">

The legacy `anon` and `service_role` keys keep working until the end of 2026. You don't have to migrate today, but doing it early lets you rotate keys safely from now on.

</Admonition>

## Before you start

The migration maps onto your existing keys:

| Legacy key     | Replace with    | Used by                                                |
| -------------- | --------------- | ------------------------------------------------------ |
| `anon`         | Publishable key | Browsers, mobile and desktop apps, CLIs, public source |
| `service_role` | Secret key      | Servers, Edge Functions, workers, other backend code   |

> For a full explanation of each key type, see [Understanding API keys](/docs/guides/getting-started/api-keys).

Reviewer: ```suggestion
For a full explanation of each key type, read [the Understanding API keys guide](/docs/guides/getting-started/api-keys).
```

## Step 1: Create the new API keys

Open the [**Settings > API Keys**](/dashboard/project/_/settings/api-keys/) section of the Dashboard and select the **Publishable and secret API keys** tab.

> Older projects don't have these keys yet. If you see a **Create new API keys** button, your project is still on legacy keys only. Creating the new keys is safe: it adds a publishable key and a secret key alongside your existing `anon` and `service_role` keys. Nothing breaks, and your legacy keys keep working.

Reviewer: ```suggestion
Older projects don't have these keys yet. If you see a **Create new API keys** button, your project is still on legacy keys only. Creating the new keys is safe. It adds a publishable key and a secret key alongside your existing `anon` and `service_role` keys. Your legacy keys keep working.
```

Again, let's not overpromise

> The new keys are created under the name `default`. You can add more keys with different names later, for example one secret key per backend component so you can rotate them independently. For an initial migration, the `default` publishable and secret keys are all you need.

Reviewer: ```suggestion
The new keys are created under the name `default`. You can add more keys with different names later, for example, one secret key per backend component, so you can rotate them independently. For an initial migration, the `default` publishable and secret keys are all you need.
```

## Step 2: Swap the publishable key in client code

Anywhere you use the `anon` key in public code, switch to the publishable key. This includes web pages, mobile and desktop apps, and any CLI or script that ships to users.

```ts
import { createClient } from '@supabase/supabase-js'

const supabase = createClient(
  'https://your-project.supabase.co',
  'sb_publishable_...' // was the anon key
)
```

> The publishable key carries the same low privileges as the `anon` key, so your [Row Level Security](/docs/guides/database/postgres/row-level-security) policies behave the same. User authentication through Supabase Auth is unchanged: the user still signs in and gets their own JWT.

Reviewer: ```suggestion
The publishable key carries the same low privileges as the `anon` key, so your [Row Level Security](/docs/guides/database/postgres/row-level-security) policies behave the same. User authentication through Supabase Auth is unchanged. The user still signs in and gets their own JWT.
```

<Admonition type="caution">

Don't hardcode a secret key in SQL or a webhook configuration, where it's stored in plain text. Store it in [Vault](/docs/guides/database/vault) and read it at call time:

```sql
headers := jsonb_build_object(
  'Content-Type', 'application/json',
  'apikey', (select decrypted_secret from vault.decrypted_secrets where name = 'secret_key')
)
```

</Admonition>

## Step 4: Update Edge Functions

Edge Functions read their keys from environment variables. Supabase adds two new ones to your functions' environment, `SUPABASE_PUBLISHABLE_KEYS` and `SUPABASE_SECRET_KEYS`, alongside the legacy `SUPABASE_ANON_KEY` and `SUPABASE_SERVICE_ROLE_KEY`. Confirm they exist in the [**Edge Functions > Secrets**](/dashboard/project/_/functions/secrets) section of the Dashboard before you start.

You have two options: a minimal change that swaps which variable you read, or a fuller upgrade to the [`@supabase/server`](https://github.com/supabase/server) SDK.

### Option 1: Read the new keys from the environment

> For most functions, the only change is how you read the key. The legacy variables held a plain string. The new ones hold a JSON object keyed by name, so you parse them and read the key by name. The key you created in [Step 1](#step-1-create-the-new-api-keys) is named `default`.

Reviewer: ```suggestion
For most functions, the only change is how you read the key. The legacy variables held a plain string. The new ones hold a JSON object keyed by name, so you parse them and read the key by name. The key you created in [step 1](#step-1-create-the-new-api-keys) is named `default`.
```

```ts
// before
const secretKey = Deno.env.get('SUPABASE_SERVICE_ROLE_KEY')!

// after
const secretKey = JSON.parse(Deno.env.get('SUPABASE_SECRET_KEYS')!)['default']
```

> The publishable keys work the same way through `SUPABASE_PUBLISHABLE_KEYS`. See [Managing Secrets](/docs/guides/functions/secrets) for more on environment variables in Edge Functions.

Reviewer: ```suggestion
The publishable keys work the same way through `SUPABASE_PUBLISHABLE_KEYS`. Read [the Managing Secrets guide](/docs/guides/functions/secrets) for more on environment variables in Edge Functions.
```

> If you created more than one secret key in [Step 1](#step-1-create-the-new-api-keys), every key lives in the same `SUPABASE_SECRET_KEYS` object, each under its own name. Read a non-default key the same way:

Reviewer: ```suggestion
If you created more than one secret key in [step 1](#step-1-create-the-new-api-keys), every key lives in the same `SUPABASE_SECRET_KEYS` object, each under its own name. Read a non-default key the same way:
```

A good way to try this is to duplicate one of your functions and migrate the copy first. See [Securing Edge Functions](/docs/guides/functions/auth) for every auth mode and use case, and [Authorization headers](/docs/guides/functions/auth-headers) for how the headers work.

## Step 5: Verify nothing uses the legacy keys

Before turning the legacy keys off, confirm nothing still depends on them. There's no automatic usage indicator, so this is a manual check. Go through every place that holds a Supabase key and make sure it now uses a publishable or secret key.

Don't forget callers that are easy to miss:

- Mobile or desktop app versions already in users' hands.
- CI/CD pipelines and deployment scripts.
- Third-party integrations and webhooks.
- Cron jobs, workers, and `pg_net` calls or Database Webhooks (see [Database Webhooks and `pg_net`](#database-webhooks-and-pgnet)).

## Step 6: Deactivate the legacy keys

> Once you've confirmed nothing uses the legacy keys, deactivate them in the [**Settings > API Keys**](/dashboard/project/_/settings/api-keys/) section. You can re-activate them if you find a client you missed, so this step is reversible.

Reviewer: ```suggestion
Once you've confirmed nothing uses the legacy keys, deactivate them in the [**Settings > API Keys**](/dashboard/project/_/settings/api-keys/) section of the Dashboard. You can re-activate them if you find a client you missed, so this step is reversible.
```
